File: monitor-switcher-wails/internal/profile/matcher.go
package profile

import (
	"monitor-switcher-wails/internal/ccd"
)

// MatchAdapterIDs updates the adapter IDs in the profile settings to match the current system.
// The adapter ID changes on system restart, so we need to match by other stable identifiers.
// This implements the same 3-tier matching strategy as the original C# code.
func MatchAdapterIDs(profile *ccd.DisplaySettings, current *ccd.DisplaySettings) {
	// Tier 1: Match by sourceInfo.id and targetInfo.id pairs
	matchByPathIDs(profile, current)

	// Tier 2 and 3 are only needed if Tier 1 fails during SetDisplayConfig
	// They are implemented in the LoadProfile function with retry logic
}

// matchByPathIDs matches adapter IDs by comparing source and target IDs.
// This is the primary matching strategy (Tier 1).
func matchByPathIDs(profile *ccd.DisplaySettings, current *ccd.DisplaySettings) {
	// Match path info adapter IDs
	for i := range profile.PathInfoArray {
		for j := range current.PathInfoArray {
			if profile.PathInfoArray[i].SourceInfo.Id == current.PathInfoArray[j].SourceInfo.Id &&
				profile.PathInfoArray[i].TargetInfo.Id == current.PathInfoArray[j].TargetInfo.Id {
				// Found matching path - update adapter IDs
				profile.PathInfoArray[i].SourceInfo.AdapterId = current.PathInfoArray[j].SourceInfo.AdapterId
				profile.PathInfoArray[i].TargetInfo.AdapterId = current.PathInfoArray[j].TargetInfo.AdapterId
				break
			}
		}
	}

	// Match mode info adapter IDs using the updated path info
	for i := range profile.ModeInfoArray {
		for j := range profile.PathInfoArray {
			if profile.ModeInfoArray[i].Id == profile.PathInfoArray[j].TargetInfo.Id &&
				profile.ModeInfoArray[i].InfoType == ccd.ModeInfoTypeTarget {
				// Found target mode - look for corresponding source mode
				for k := range profile.ModeInfoArray {
					if profile.ModeInfoArray[k].Id == profile.PathInfoArray[j].SourceInfo.Id &&
						profile.ModeInfoArray[k].AdapterId.LowPart == profile.ModeInfoArray[i].AdapterId.LowPart &&
						profile.ModeInfoArray[k].InfoType == ccd.ModeInfoTypeSource {
						// Update source mode adapter ID
						profile.ModeInfoArray[k].AdapterId = profile.PathInfoArray[j].SourceInfo.AdapterId
						break
					}
				}
				// Update target mode adapter ID
				profile.ModeInfoArray[i].AdapterId = profile.PathInfoArray[j].TargetInfo.AdapterId
				break
			}
		}
	}
}

// MatchByMonitorName matches adapter IDs by comparing monitor friendly names.
// This is the fallback strategy (Tier 2) used when Tier 1 fails.
func MatchByMonitorName(profile *ccd.DisplaySettings, current *ccd.DisplaySettings) {
	for i := range profile.ModeInfoArray {
		for j := range current.AdditionalInfo {
			if current.AdditionalInfo[j].MonitorFriendlyDevice != "" &&
				i < len(profile.AdditionalInfo) &&
				profile.AdditionalInfo[i].MonitorFriendlyDevice != "" &&
				current.AdditionalInfo[j].MonitorFriendlyDevice == profile.AdditionalInfo[i].MonitorFriendlyDevice {

				originalID := profile.ModeInfoArray[i].AdapterId

				// Update all path info with matching adapter ID
				for k := range profile.PathInfoArray {
					if profile.PathInfoArray[k].TargetInfo.AdapterId == originalID {
						profile.PathInfoArray[k].TargetInfo.AdapterId = current.ModeInfoArray[j].AdapterId
						profile.PathInfoArray[k].SourceInfo.AdapterId = current.ModeInfoArray[j].AdapterId
						profile.PathInfoArray[k].TargetInfo.Id = current.ModeInfoArray[j].Id
					}
				}

				// Update all mode info with matching adapter ID
				for k := range profile.ModeInfoArray {
					if profile.ModeInfoArray[k].AdapterId == originalID {
						profile.ModeInfoArray[k].AdapterId = current.ModeInfoArray[j].AdapterId
					}
				}

				profile.ModeInfoArray[i].AdapterId = current.ModeInfoArray[j].AdapterId
				profile.ModeInfoArray[i].Id = current.ModeInfoArray[j].Id
				break
			}
		}
	}
}

// MatchByBulkReplacement replaces all instances of an old adapter ID with a new one.
// This is the last resort strategy (Tier 3).
func MatchByBulkReplacement(profile *ccd.DisplaySettings, current *ccd.DisplaySettings) {
	for i := range profile.PathInfoArray {
		for j := range current.PathInfoArray {
			if profile.PathInfoArray[i].SourceInfo.Id == current.PathInfoArray[j].SourceInfo.Id &&
				profile.PathInfoArray[i].TargetInfo.Id == current.PathInfoArray[j].TargetInfo.Id {

				oldID := profile.PathInfoArray[i].SourceInfo.AdapterId.LowPart
				newID := current.PathInfoArray[j].SourceInfo.AdapterId.LowPart

				// Replace all occurrences in path info
				for k := range profile.PathInfoArray {
					if profile.PathInfoArray[k].SourceInfo.AdapterId.LowPart == oldID {
						profile.PathInfoArray[k].SourceInfo.AdapterId.LowPart = newID
					}
					if profile.PathInfoArray[k].TargetInfo.AdapterId.LowPart == oldID {
						profile.PathInfoArray[k].TargetInfo.AdapterId.LowPart = newID
					}
				}

				// Replace all occurrences in mode info
				for k := range profile.ModeInfoArray {
					if profile.ModeInfoArray[k].AdapterId.LowPart == oldID {
						profile.ModeInfoArray[k].AdapterId.LowPart = newID
					}
				}
				break
			}
		}
	}
}

File: monitor-switcher-tauri/src-tauri/src/ccd/matcher.rs
//! Adapter ID matching logic for display profiles.
//!
//! Adapter IDs (LUIDs) change on system restart, so we need to match profiles
//! to current system state using multiple fallback strategies.

use super::types::*;
use super::api::{DisplaySettings, MonitorAdditionalInfo, get_display_settings, get_monitor_additional_info};
use log::{debug, warn};

/// Match profile adapter IDs to current system adapter IDs.
/// Uses a 3-tier fallback strategy:
/// 1. Match by source/target ID pairs
/// 2. Match by monitor friendly name (EDID)
/// 3. Bulk adapter ID replacement
pub fn match_adapter_ids(
    settings: &mut DisplaySettings,
    additional_info: &[MonitorAdditionalInfo],
) -> Result<(), String> {
    // Get current display settings
    let current = get_display_settings(true)?;
    let current_additional_info = get_additional_info_for_modes(&current.mode_info_array);

    // Try tier 1: Match by source/target ID pairs
    if try_match_by_ids(settings, &current) {
        debug!("Adapter matching: Tier 1 (ID pairs) succeeded");
        return Ok(());
    }

    // Try tier 2: Match by monitor friendly name
    if try_match_by_friendly_name(settings, additional_info, &current, &current_additional_info) {
        debug!("Adapter matching: Tier 2 (friendly name) succeeded");
        return Ok(());
    }

    // Try tier 3: Bulk replacement
    if try_bulk_replacement(settings, &current) {
        debug!("Adapter matching: Tier 3 (bulk replacement) succeeded");
        return Ok(());
    }

    warn!("Adapter matching: All tiers failed, using original IDs");
    Ok(())
}

/// Tier 1: Match by source and target ID pairs.
fn try_match_by_ids(settings: &mut DisplaySettings, current: &DisplaySettings) -> bool {
    let mut matched_any = false;

    // Match paths by source/target IDs
    for path in &mut settings.path_info_array {
        for current_path in &current.path_info_array {
            if path.source_info.id == current_path.source_info.id
                && path.target_info.id == current_path.target_info.id
            {
                path.source_info.adapter_id = current_path.source_info.adapter_id;
                path.target_info.adapter_id = current_path.target_info.adapter_id;
                matched_any = true;
                break;
            }
        }
    }

    if !matched_any {
        return false;
    }

    // Match mode infos by correlating with paths
    for mode in &mut settings.mode_info_array {
        // Find a path that references this mode's adapter
        for path in &settings.path_info_array {
            if mode.info_type == MODE_INFO_TYPE_TARGET && mode.id == path.target_info.id {
                // Find current mode with same id
                for current_mode in &current.mode_info_array {
                    if current_mode.info_type == MODE_INFO_TYPE_TARGET
                        && current_mode.id == mode.id
                    {
                        mode.adapter_id = current_mode.adapter_id;
                        break;
                    }
                }
                break;
            } else if mode.info_type == MODE_INFO_TYPE_SOURCE && mode.id == path.source_info.id {
                for current_mode in &current.mode_info_array {
                    if current_mode.info_type == MODE_INFO_TYPE_SOURCE
                        && current_mode.id == mode.id
                    {
                        mode.adapter_id = current_mode.adapter_id;
                        break;
                    }
                }
                break;
            }
        }
    }

    true
}

/// Tier 2: Match by monitor friendly device name.
fn try_match_by_friendly_name(
    settings: &mut DisplaySettings,
    additional_info: &[MonitorAdditionalInfo],
    current: &DisplaySettings,
    current_additional_info: &[MonitorAdditionalInfo],
) -> bool {
    let mut matched_any = false;

    for (i, mode) in settings.mode_info_array.iter_mut().enumerate() {
        if mode.info_type != MODE_INFO_TYPE_TARGET {
            continue;
        }

        let Some(saved_info) = additional_info.get(i).filter(|info| info.valid) else {
            continue;
        };
        if saved_info.monitor_friendly_device.is_empty() {
            continue;
        }

        // Find matching current monitor by friendly name
        for (j, current_mode) in current.mode_info_array.iter().enumerate() {
            if current_mode.info_type != MODE_INFO_TYPE_TARGET {
                continue;
            }

            let Some(current_info) = current_additional_info.get(j).filter(|info| info.valid) else {
                continue;
            };

            if current_info.monitor_friendly_device == saved_info.monitor_friendly_device {
                mode.adapter_id = current_mode.adapter_id;
                mode.id = current_mode.id;
                matched_any = true;
                break;
            }
        }
    }

    if matched_any {
        // Update paths based on matched modes
        update_path_adapter_ids_from_modes(settings, current);
    }

    matched_any
}

/// Tier 3: Bulk replacement of old adapter IDs with new ones.
fn try_bulk_replacement(settings: &mut DisplaySettings, current: &DisplaySettings) -> bool {
    // Find one matching path to get the old->new adapter ID mapping
    for path in &settings.path_info_array {
        for current_path in &current.path_info_array {
            // Try to find any matching criteria
            if path.source_info.id == current_path.source_info.id {
                let old_id = path.source_info.adapter_id;
                let new_id = current_path.source_info.adapter_id;

                if old_id != new_id {
                    replace_all_adapter_ids(settings, old_id, new_id);
                    return true;
                }
            }
        }
    }

    false
}

/// Replace all occurrences of old adapter ID with new one.
fn replace_all_adapter_ids(settings: &mut DisplaySettings, old_id: LUID, new_id: LUID) {
    for path in &mut settings.path_info_array {
        if path.source_info.adapter_id == old_id {
            path.source_info.adapter_id = new_id;
        }
        if path.target_info.adapter_id == old_id {
            path.target_info.adapter_id = new_id;
        }
    }

    for mode in &mut settings.mode_info_array {
        if mode.adapter_id == old_id {
            mode.adapter_id = new_id;
        }
    }
}

/// Update path adapter IDs based on matched mode adapter IDs.
fn update_path_adapter_ids_from_modes(settings: &mut DisplaySettings, current: &DisplaySettings) {
    for path in &mut settings.path_info_array {
        // Find current path with same source/target IDs if possible
        for current_path in &current.path_info_array {
            if path.source_info.id == current_path.source_info.id {
                path.source_info.adapter_id = current_path.source_info.adapter_id;
            }
            if path.target_info.id == current_path.target_info.id {
                path.target_info.adapter_id = current_path.target_info.adapter_id;
            }
        }
    }
}

/// Get additional info for all target modes in the array.
pub fn get_additional_info_for_modes(mode_info_array: &[DisplayConfigModeInfo]) -> Vec<MonitorAdditionalInfo> {
    mode_info_array
        .iter()
        .map(|mode| {
            if mode.info_type == MODE_INFO_TYPE_TARGET {
                get_monitor_additional_info(mode.adapter_id, mode.id)
            } else {
                MonitorAdditionalInfo::default()
            }
        })
        .collect()
}
